lte/gateway/c/session_manager/LocalEnforcer.cpp

```diff
@@ -442,68 +442,91 @@ static RedirectInformation_AddressType address_type_converter(
   }
 }
 
-static PolicyRule create_redirect_rule(
-    const std::unique_ptr<ServiceAction>& action) {
+PolicyRule LocalEnforcer::create_redirect_rule(
+    const RedirectInstallInfo& info) {
   PolicyRule redirect_rule;
   redirect_rule.set_id("redirect");
   redirect_rule.set_priority(LocalEnforcer::REDIRECT_FLOW_PRIORITY);
-  action->get_credit_key().set_rule(&redirect_rule);
 
   RedirectInformation* redirect_info = redirect_rule.mutable_redirect();
   redirect_info->set_support(RedirectInformation_Support_ENABLED);
 
-  auto redirect_server = action->get_redirect_server();
+  auto redirect_server = info.redirect_server;
   redirect_info->set_address_type(
       address_type_converter(redirect_server.redirect_address_type()));
   redirect_info->set_server_address(redirect_server.redirect_server_address());
 
   return redirect_rule;
 }
 
-void LocalEnforcer::install_redirect_flow(
-    const std::unique_ptr<ServiceAction>& action,
+void LocalEnforcer::start_redirect_flow_install(
+    SessionMap& session_map, const std::unique_ptr<ServiceAction>& action_p,
     SessionUpdate& session_update) {
+  // Bundle up all info into this struct so that we don't have to pass around
+  // unique pointers
+  RedirectInstallInfo redirect_info{
+      .imsi            = action_p->get_imsi(),
+      .session_id      = action_p->get_session_id(),
+      .redirect_server = action_p->get_redirect_server(),
+  };
+
+  MLOG(MDEBUG) << "Fetching Subscriber IP address from DirectoryD for "
+               << redirect_info.session_id;
+  directoryd_client_->get_directoryd_ip_field(
+      redirect_info.imsi,
+      [this, redirect_info](Status status, DirectoryField resp) {
+        // This call back gets executed in the DirectoryD client thread, but
+        // we want to run the session update logic in the main thread.
+        evb_->runInEventBaseThread([this, redirect_info, status, resp]() {
```

Comment:
what is the implication of delaying the redirection? Do we have to provide any answer to the core that the redirection was done? I understand that while redirection is not done the user will enjoy of some free data>? (which would be fine taking into consideration the lack of real events on sessiond)

Reply:
@uri200  This is not a delay, but it's more that we want to run all of the logic that touches SessionState/SessionStore in the main thread. This callback function is passed into the DirectoryD client, which means that it's run in DirectoryD's response handling thread. (Different from the main thread) I was wondering if we were seeing any race condition in accessing the session store with these two separate threads accessing the store. So this new logic just moves the handling of the received IP into the main thread.

Reply:
Oh, maybe I missed your question. :) The delay should not be that great, as it's only a GRPC exchange with a local DirectoryD service. The redirection install is not triggered by a message from the core, but quota exhaustion. So I suppose waiting for the directoryd response will extend the "free internet" period by a bit, but we already have a 2second error margin due to PipelineD's reporting period anyways.